babel_monitor/src/lib.rs

```diff
@@ -131,57 +97,81 @@ impl<T: Read + Write> InnerBabel<T> {
             Ok(ret)
         } else {
-            Err(Error::BabelError(ret))
+            Err(CommandFailed(ret).into())
         }
     }
 
-    fn write(&mut self, command: &str) -> Result<(), Error> {
-        self.stream.get_mut().write(command.as_bytes())?;
-        trace!("sending {} to babel", command);
+    fn write_babel(&mut self, command: &str) -> Result<(), Error> {
+        self.write_all(command.as_bytes())?;
+        trace!("sent {} to babel", command);
         Ok(())
     }
+}
+
+impl Babel {
+    // Consumes the automated Preamble and validates configuration api version
+    pub fn start_connection(&mut self) -> Result<(), Error> {
+        trace!("About to get the preamble");
+        let preamble = self.read_babel()?;
+        trace!("Got the preamble: {}", preamble);
+        // Note you have changed the config interface, bump to 1.1 in babel
+        if preamble.contains("ALTHEA 0.1") && positive_termination(&preamble) {
+            info!("Attached OK to Babel with preamble: {}", preamble);
+            return Ok(());
+        } else {
+            return Err(InvalidPreamble(preamble).into());
+        }
+    }
 
     pub fn local_fee(&mut self) -> Result<u32, Error> {
-        self.write("dump\n")?;
-        for entry in self.read()?.split("\n") {
-            if entry.contains("local fee") {
-                return Ok(find_babel_val("fee", entry)?.parse()?);
-            }
+        self.write_babel("dump\n")?;
+
+        let babel_output = self.read_babel()?;
+        let fee_entry = match babel_output.split("\n").nth(0) {
+            Some(entry) => entry,
+            // Even an empty string wouldn't yield None
+            None => return Err(LocalFeeNotFound(String::from("<Babel output is None>")).into()),
+        };
+
+        if fee_entry.contains("local fee") {
+            let fee = find_babel_val("fee", fee_entry)?.parse()?;
+            trace!("Retrieved a local fee of {}", fee);
+            return Ok(fee);
         }
-        Ok(0)
+
+        Err(LocalFeeNotFound(String::from(fee_entry)).into())
     }
 
-    pub fn monitor(&mut self, iface: &str) -> Result<u32, Error> {
+    pub fn monitor(&mut self, iface: &str) -> Result<(), Error> {
         let commmand = format!("interface {} \n", iface);
-        self.write(&commmand)?;
-        let out = self.read()?;
+        self.write_babel(&commmand)?;
+        let _ = self.read_babel()?;
         info!("Babel started monitoring: {}", iface);
-        Ok(0)
+        Ok(())
     }
 
-    pub fn redistribute_ip(&mut self, ip: &IpAddr, allow: bool) -> Result<u32, Error> {
+    pub fn redistribute_ip(&mut self, ip: &IpAddr, allow: bool) -> Result<(), Error> {
         let commmand = format!(
             "redistribute ip {}/128 {}\n",
             ip,
             if allow { "allow" } else { "deny" }
         );
-        self.write(&commmand)?;
-        let out = self.read()?;
-        Ok(0)
+        self.write_babel(&commmand)?;
+        let _ = self.read_babel()?;
```

Comment:
we aren't reading the command outputs of babel, so how do we know if any of these succeeded?